src/functional_tests.rs

```
@@ -761,6 +761,302 @@ mod handle_malice {
+        let b_0 = unwrap!(nth_event(bob.graph(), 0).pack(bob.event_context()));
+        let b_1 = unwrap!(nth_event(bob.graph(), 1).pack(bob.event_context()));
+        let a_0 = unwrap!(nth_event(bob.graph(), 2).pack(bob.event_context()));
+        let a_1 = unwrap!(nth_event(bob.graph(), 3).pack(bob.event_context()));
+        let alice_requesting = unwrap!(nth_event(bob.graph(), 4).pack(bob.event_context()));
+        let invalid_req = PackedEvent::new_request(
+            bob.our_pub_id().clone(),
+            b_1.compute_hash(),
+            alice_requesting_hash,
+        );
+        let request = Request {
+            packed_events: vec![b_0, b_1, a_0, a_1, alice_requesting, invalid_req],
+        };
+
+        assert_matches!(
+            alice.handle_request(bob.our_pub_id(), request.clone()),
+            Err(Error::InvalidEvent)
+        );
+
+        assert_matches!(
+            dave.handle_request(bob.our_pub_id(), request),
+            Err(Error::InvalidEvent)
+        );
+    }
+
+    #[test]
+    fn invalid_request_wrong_type_of_other_parent() {
+        let (mut alice, mut bob, mut carol) =
+            unwrap!(initialise_genesis_parsecs(3).into_iter().collect_tuple());
+
+        // Send request from Alice to Bob.
+        let request = unwrap!(alice.create_gossip(bob.our_pub_id()));
+        let alice_requesting_hash = *nth_event(alice.graph(), 2).hash();
+
+        assert!(bob.handle_request(alice.our_pub_id(), request).is_ok());
+
+        // Create invalid request from Bob.
+        let b_0 = unwrap!(nth_event(bob.graph(), 0).pack(bob.event_context()));
+        let b_1 = unwrap!(nth_event(bob.graph(), 1).pack(bob.event_context()));
+        let a_0 = unwrap!(nth_event(bob.graph(), 2).pack(bob.event_context()));
+        let a_1 = unwrap!(nth_event(bob.graph(), 3).pack(bob.event_context()));
+        let alice_requesting = unwrap!(nth_event(bob.graph(), 4).pack(bob.event_context()));
+        let invalid_req = PackedEvent::new_request(
```

This would actually be a valid `Request` event.  The only reason the test is passing is that we call `alice.handle_request()` below; we should be calling `alice.handle_response()`.  If we passed that current message (as a response message) to alice, it would be accepted and the test would fail.

---

The task description says
"3, Call Alice unpack_and_add_events(vec![Bob's Request]) - should fail."
You mean we shall actually use `alice.handle_response` to accept this, and still expect handle_response return with an error?
or you mean if we do so, handle_response returns ok, which indicates a bug?

---

I think it should be fine to use `handle_response()` for this purpose here - it should return an error if the response message does actually contain an invalid `Request` event created by Bob.  (The one currently being created is valid, but if you make the change as JP suggested it won't be.)  That invalid event shouldn't get added to anyone's graph, and accusations should be made.

However, all this is true if we just call `unpack_and_add_events()` too.  As long as we're checking for invalid events not being added to graphs and being added to appropriate accusations, I don't think there's much difference between either option.

---

first, I'd prefer not to open another function as public.
second, I'd like this set of tests to simulate the normal usage (i.e. use handle_response/handle_request only), to give an example how requestting-request-response pattern works.

Calling `unpack_and_add_events()` will hide the pattern a lot to me.

---

I'm fine with either approach, as long as sticking to just the existing public API of `TestParsec` doesn't introduce a lot of extra code to the tests.  But I doubt that'll be the case here; so go with whatever you prefer :)